### PR TITLE
[Feature] Adding quantum information tools

### PR DESCRIPTION
**Pull request by inafergra**
This PR adds two functionalities related to quantum information:

- Quantum Fisher Information (QFI)
- Quantum Natural Gradient (QNG): gradient-optimizer that uses the QFI matrix to guide the parameter update. It can either be used with the "exact" QFI or using a SPSA approximation (https://arxiv.org/pdf/2103.09232.pdf)

The code works but it is not very efficient, in particular the QNG training loop. Most of the code was developed for previous versions of qadence so there could also more straightforward ways to calculate the QFI matrix. Any help optimizing the code and making the code more efficient is very welcome!

### CODE REVIEW

File: qadence/qinfo_tools/qfi.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+from qadence import DiffMode
+from qadence.blocks import primitive_blocks, parameters
+
+
+def symsqrt(a, cond=None, return_rank=False):
+    """Computes the symmetric square root of a positive definite matrix.
+    Taken from https://github.com/pytorch/pytorch/issues/25481"""
+
+    s, u = torch.linalg.eigh(a)
+    cond_dict = {
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:19]: I guess thus can be an Enum instead.

File: qadence/qinfo_tools/qfi.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+from qadence.circuit import QuantumCircuit
+from qadence.backend import BackendName
+from qadence.measurements import Measurements
+from qadence import DiffMode
+from qadence.blocks import primitive_blocks, parameters
+
+
+def symsqrt(a, cond=None, return_rank=False):
+    """Computes the symmetric square root of a positive definite matrix.
+    Taken from https://github.com/pytorch/pytorch/issues/25481"""
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:16]: It is fair to cite the origin of the code.

File: qadence/qinfo_tools/qfi.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    if return_rank:
+        return B, len(psigma_diag)
+    else:
+        return B
+
+
+def _overlap_with_inputs(
+    circuit: QuantumCircuit,
+    vparams: tuple | None = None,
+    backend: BackendName = BackendName.PYQTORCH,
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:42]: Have you tested what is the performance with the other backends and GPSR? I am asking out of curiousity.

File: qadence/qinfo_tools/qfi.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+    overlap_method: OverlapMethod = OverlapMethod.EXACT,
+    diff_mode: DiffMode = DiffMode.AD,
+):
+    """Gives an OverlapModel consisting of the overlap of the circuit with itself. The vparamaters
+    of the underlying QuantumModel are the vparameters of the bra state, so it's differentiable.
+
+    Args:
+        circuit (QuantumCircuit): _description_
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:50]: I guess the docstring here was WIP by definition.

File: qadence/qinfo_tools/qng.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+            else:
+                raise NotImplementedError(
+                    "Non valid approximation, please choose 'exact' or 'spsa'."
+                )
+
+        return loss
+
+
+# class QuantumNaturalGradient(Optimizer):
```
Comment (gvelikova) [inline qadence/qinfo_tools/qng.py:102]: The commented code can be cleaned out.

File: qadence/qinfo_tools/qfi.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+        qfi_mat_estimator = (1 / (k + 1)) * (k * previous_qfi_estimator + qfi_mat)
+
+    # Get the positive-semidefinite version of the matrix for the update rule in QNG
+    qfi_mat_positive_sd = symsqrt(torch.matmul(qfi_mat_estimator, qfi_mat_estimator))
+    qfi_mat_positive_sd = qfi_mat_positive_sd + beta * torch.eye(num_vparams)
+
+    return qfi_mat_estimator, qfi_mat_positive_sd
+
+
+# def symsqrt(a, cond=None, return_rank=False):
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:252]: The commented code can be cleaned out.

**Comment by gvelikova** [comment]
@inafergra thank you for adding the QNG and QFI  as qadence features.

File: qadence/qinfo_tools/qfi.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            + torch.matmul(Delta_2, Delta_1.transpose(0, 1))
+        )
+    )
+    qfi_mat = -2 * fid_hess
+
+    # Calculate the new estimator from the old estimator of qfi_mat
+    if k == 0:
+        qfi_mat_estimator = qfi_mat
+    else:
+        qfi_mat_estimator = (1 / (k + 1)) * (k * previous_qfi_estimator + qfi_mat)  # type: ignore
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:255]: This previous_qfi_estimator isn't defined anywhere it seems.

File: qadence/qinfo_tools/qfi.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    Args:
+        circuit (QuantumCircuit): _description_
+        backend (BackendName, optional): _description_. Defaults to BackendName.pyq.
+        overlap_method (OverlapMethod, optional): _description_. Defaults to OverlapMethod.EXACT.
+        diff_mode (DiffMode, optional): _description_. Defaults to DiffMode.ad.
+
+    Returns:
+        torch.Tensor: QFI matrix
+    """
+    num_vparams = len(var_values) if var_values is not None else 0
```
Comment (gvelikova) [inline qadence/qinfo_tools/qfi.py:186]: I added the if/else condition depending on type. When would you have None parameters and should it make sense to add a better checking?

**Comment by dominikandreasseitz** [comment]
thanks a lot @inafergra , had a quick glance and it looks very nice. good job! i'll have a closer look sometime tomorrow or friday

File: qadence/qinfo_tools/qfi.py (hunk 3 above)
Comment (inafergra) [reply qadence/qinfo_tools/qfi.py:44]: Not yet! I have only tested pyqtorch for the moment

File: qadence/qinfo_tools/qfi.py (hunk 7 above)
Comment (inafergra) [reply qadence/qinfo_tools/qfi.py:255]: True! I think the docstrings need to be updated in general, I will check them all and make sure they are updated.

`previous_qfi_estimator` is an input to the `get_quantum_fisher_spsa()`. This function computes the estimator of the QFI matrix corresponding to the iteration k (also an input to the function). You need the estimator from the previous iteration (k-1) to construct the new estimator.

File: qadence/qinfo_tools/qfi.py (hunk 8 above)
Comment (inafergra) [reply qadence/qinfo_tools/qfi.py:186]: `var_values` are the values of the variational parameters where you want to compute your QFI matrix. When they are set to `None`, the QFI is calculated in with the default values for the vparams in the model (for example, if you call it after training your model and you want to compute the QFI with the values of the optimal vparams, you can just set `var_values` to None and it will automatically compute it there). However, if you want to calculate the QFI in another point of your landscape, you can pass those values in the input.

Comment (gvelikova) [reply qadence/qinfo_tools/qfi.py:186]: OK, thanks for the explanation, but the original len(None) in these cases wouldn't work or am I missing something?

File: examples/qinfo_tools/quantum_natural_gradient.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+# Plot losses
+fig, _ = plt.subplots()
+plt.plot(range(n_epochs_adam), loss_adam, label="Adam optimizer")
+plt.plot(range(n_epochs_qng), loss_qng, label="QNG optimizer")
+plt.plot(range(n_epochs_qng_spsa), loss_qng_spsa, label="QNG-SPSA optimizer")
+plt.legend()
+plt.xlabel("Training epochs")
+plt.ylabel("Loss")
+plt.show()
```
Comment (dominikandreasseitz) [inline examples/qinfo_tools/quantum_natural_gradient.py:124]: lets make a tutorial in markdown format too. we can show plots there. however we shouldnt plot in example files since they are run in the CI

**Comment by Roland-djee** [comment]
@inafergra As you are probably aware, there is now [qadence-libs](https://github.com/pasqal-io/qadence-libs) where this work is best suited. Thanks for the contribution!

**Comment by inafergra** [comment]
> @inafergra As you are probably aware, there is now [qadence-libs](https://github.com/pasqal-io/qadence-libs) where this work is best suited. Thanks for the contribution!

**Comment by inafergra** [comment]
Closing the PR due to move to [qadence-libs](https://github.com/pasqal-io/qadence-libs)
